add chord function and update docs

timmydoza opened the pull request.

dingram commented on README.md:
```diff
@@ -86,29 +90,32 @@ and the use of `addNoteOn()`/`addNoteOff()` to produce chords.
    - If `velocity` is given, strike the note with that velocity
+ - `addChord(channel, chord[, velocity])`
+
+   **Add a chord with the given channel and pitches**
```
I would add a bullet about the format of the `chord` argument.

dingram commented on lib/jsmidgen.js:
```diff
@@ -485,6 +485,31 @@ var Midi = {};
+	 * @returns {Track} The current track.
+	 */
+	Track.prototype.addChord = Track.prototype.chord = function(channel, chord, dur, velocity) {
+		if (!Array.isArray(chord)) {
```
Also to be ultra-defensive, please check the array has at least one element.

walmik commented on lib/jsmidgen.js:
```diff
@@ -485,6 +485,31 @@ var Midi = {};
+			throw new Error('Chord must be an array of pitches');
+		}
+		for (var i = 0; i < chord.length; i++) {
+			this.noteOn(channel, chord[i], velocity);
```
I m off the understanding that this function will need to have the third argument only for the first note. If you add this for each note, then it may result in an unexpected outcome in the MIDI file itself. It d be good to add a test or confirm with @dingram about this.

dingram replied:
Good catch. There's a missing `0, ` before the `velocity` argument.

walmik commented on lib/jsmidgen.js:
```diff
@@ -485,6 +485,31 @@ var Midi = {};
+			this.noteOn(channel, chord[i], velocity);
+		}
+		this.noteOff(0, chord[0], dur);
+		for (var i = 1; i < chord.length; i++) {
```
Can we please use `forEach` instead of a `for` loop?

timmydoza commented:
Thanks for your comments.  I added a bullet about the format of the chord argument in the readme.  I also corrected the issue of the missing duration argument. 

I could use forEach instead of the first for loop if you want.  The second for loop iterates over all pitches except for the first, so forEach wouldn't work for that one.

walmik commented:
Hello Timmy, the second param in a `forEach` is an index of the item currently in the loop, hence you can use it.

```
chord.forEach(function(el, idx) {
    if (idx === 0) {
        // something that needs to be done only on the first element
    }
    // affect all elements that come here
});
```

Hope this helps :)

timmydoza commented:
Thanks walmik.  I had that idea.  I'm new to this, so I'm not exactly sure why forEach is superior to a for loop in this instance.  Is it that forEach does a better job of expressing the intent of the code?  I appreciate your help.

walmik commented:
Hi Timmy,

Yes, `forEach` (or `map`) aids in expressing the intent and there s more. Most JS developers have switched to `forEach` and `map` for some time now. I ll try to give some background for this switch that took place a while ago, hopefully it ll help.

Speed wise a basic `for` loop is faster but that s about it. In this specific repo at the code where we are wondering whether we should use for or forEach, speed is not a constraint. What we are then left with is readability, maintainability and maybe scalability. A `forEach` or a `map` is much more readable (and shorter in terms of characters). It takes less memory too.

For a more broader discussion on why `forEach` or `map` instead of `for`, you might be aware that ES5 has function scope (no block scope). This may not be directly related to the codebase we are contributing to, but lets assume that we had some async action that we needed to do on the `chord` array. For eg try running this sample code in your browser's console:

```
var chord = ["c3", "e3", "g3"];

chord.forEach(function(element) {
  setTimeout(function() {
    console.log(element);
  }, 100);
});


for (var i = 0; i < chord.length; i++) {
  setTimeout(function() {
     console.log(chord[i]);
  }, 100);
}
```

You ll notice that the for loop just ended up throwing out `undefined` 3 times. Of course you can fix that be using a closure and binding the value of i to it but that ll only add up more cruft to the code making it less readable and intuitive. 

Another example is, we can define a function and use that within the forEach as well as elsewhere making it more scalable. There are more gotchas and you can search for them .. Also, there are a coupla `for` loops in this code base as well and maybe we can clean that up later. But it d be nice to follow forEach (or map) going forward if @dingram says that as well :)

timmydoza commented:
Thanks for your help walmik!  Your comments make sense and they are super helpful.

I've replaced the for loops with forEach calls.  I had to use the bind method so that the correct 'this' was accessible inside the forEach call.  Does this look ok to you, or can you think of any alternatives to using bind?  Thanks again.

dingram commented:
The only alternative to `bind()` that I can think of would be capturing `this` beforehand:

```
  var thisTrack = this;
  chord.forEach(function(note) {
    thisTrack.noteOn(channel, note, 0, velocity);
  });
```

walmik commented:
Well, `forEach` to the rescue again, `forEach` allows for a yet another param to pass as the value of `this` while executing the fallback. Which means we dont need `bind` nor do we need to cache the value of `this`. 

```
arr.forEach(callback[, thisArg])
```

So `callback` is the first argument and it takes `element`, `index` and the `array` itself (we dont really need to use this one) and the second argument is `thisArg` where you can pass `this` to denote Track. 

@timmydoza Can you please try this and share with us if that works?

walmik commented:
I m also of the opinion (and maybe you can verify this, that the noteOn method also needs the index check. So putting it all together it looks like this. It ll be nice if you can try it and test it thoroughly form your side by actually opening up the MIDI files this generated and importing them in whatever music software you use to check if the output was indeed as expected:

```
Track.prototype.addChord = Track.prototype.chord = function(channel, chord, dur, velocity) {
	if (!Array.isArray(chord) && !chord.length) {
		throw new Error('Chord must be an array of pitches');
	}
	chord.forEach(function(note) {
		if (index === 0) {
			this.noteOn(channel, note, 0, velocity);
		} else {
			this.noteOn(channel, note);
		}
	}, this);
	chord.forEach(function(note, index) {
		if (index === 0) {
			this.noteOff(channel, note, dur, velocity);
		} else {
			this.noteOff(channel, note);
		}
	}, this);
	return this;
};
```

timmydoza commented:
Thanks again walmik!  I was able to get rid of the bind call by using 'this' as an argument in the forEach function as you suggested.  Good call!!

I'm not sure I understand what you are going for with your second suggestion.  I think you may have accidentally omitted the velocity argument from the second noteOn call.  Also I think you maybe meant to pass the 'dur' variable to the first noteOn call instead of 0.  Not sure.

Adding the 'dur' argument to the very first noteOn call means that the code below will give you three chords of duration 64 with a silence of duration 64 between each chord, instead of three chords of duration 64 played in immediate succession.  

```javascipt
track.addChord(0, ['c4', 'e4', 'g4'], 64);
track.addChord(0, ['d4', 'f4', 'a4'], 64);
track.addChord(0, ['e4', 'g4', 'b4'], 64);
```

I'd love to hear your thoughts on the matter.  Perhaps I'm missing something.  I've manually tested the chord function as it stands right now and it works.

walmik commented:
You are very welcome :)

If you have tested and are able to confirm then that s great. I had implemented something similar in a project where I use jsmidgen as a dependency and there I had to explicitly check index even in case of noteOn.

From my side I m good with the changes.
